### PR TITLE
DOCSP-6562 FLE guide data keys

### PR DESCRIPTION
**Pull request by ccho-mongodb**
JIRA: https://jira.mongodb.org/browse/DOCSP-6562

Staging: https://docs-mongodborg-staging.corp.mongodb.com/ecosystem/docsworker/fle-guide-data-keys/use-cases/sensitive-data-encryption.html

### CODE REVIEW

File: source/use-cases/sensitive-data-encryption.txt [hunk 1]
```diff
@@ -124,6 +124,151 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+* A MongoDB connection string that authenticates to a running server.
+* The key vault database and collection name.
+
+Upon creating a data encryption key, the driver returns the ``id`` field of
+a document that stores information about the encrypted key including:
+
+* The data encryption key, encrypted by the master key.
+* The key management service provider data.
+
+In this section, we generate a single data encryption key from the local
```
Comment (terakilobyte) [inline source/use-cases/sensitive-data-encryption.txt]: "In this section" is used two paragraphs above. Suggest rewording, but not a blocker.

File: source/use-cases/sensitive-data-encryption.txt [hunk 2]
```diff
@@ -124,6 +124,151 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+In this section, we use the MongoDB driver to create and store a data 
+encryption key, used by CSFLE to automatically encrypt and decrypt data on 
+document fields. The following configuration values are required by the client 
+to generate a new data encryption key:
+
+* The local master key or AWS KMS master key access settings.
+* A MongoDB connection string that authenticates to a running server.
+* The key vault database and collection name.
+
+Upon creating a data encryption key, the driver returns the ``id`` field of
```
Comment (terakilobyte) [inline source/use-cases/sensitive-data-encryption.txt]: Shouldn't it be ``_id`` instead of ``id``? It also might be worth adding an admonition that the _id field of a key is *not* an ObjectID.

File: source/use-cases/sensitive-data-encryption.txt (hunk 1 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Thanks for identifying that, will reword.

File: source/use-cases/sensitive-data-encryption.txt (hunk 2 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: I agree, I should use `_id`. Should I identify this as a `UUID` for the reader or is that too much detail?

Comment (terakilobyte) [reply source/use-cases/sensitive-data-encryption.txt]: I think you should call it out for sure. It will be surprising that is a UUID instead of an ObjectID.

Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Fixed, let me know if it looks ok now.

File: source/use-cases/sensitive-data-encryption.txt [hunk 3]
```diff
@@ -124,7 +124,153 @@ In this step, we generate a local master key to expedite setup of our developmen
           }
 
 
-B. Define a JSON Schema
+B. Create a Data Encryption Key
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In this section, we use the MongoDB driver to create and store a data
+encryption key, used by CSFLE to automatically encrypt and decrypt data on
+document fields. The following configuration values are required by the client
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: nit: IMO the comma before "used" makes it read like an appositive. Instead we could do two sentences:

In this section we use the MongoDB driver to create and store a data encryption key _in the keyvault collection_. Drivers that have access to the keyvault can use the encryption key to encrypt and decrypt fields in a document.

Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: nit: passive -> active

The driver requires the following configuration values to generate a new data encryption key:

File: source/use-cases/sensitive-data-encryption.txt [hunk 4]
```diff
@@ -124,7 +124,153 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+to generate a new data encryption key:
+
+* The local master key or AWS KMS master key access settings.
+* A MongoDB connection string that authenticates on a running server.
+* The key vault database and collection name.
+
+Upon creating a data encryption key, the driver returns the ``_id`` field
+(as a `UUID <https://en.wikipedia.org/wiki/Universally_unique_identifier>`_)
+of a document that stores information about the key including:
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: Are we certain that this is a UUID? I would imagine that it's a 12-byte ObjectId value unless this specifically does not use the default.

Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: I might interpret this as "the _id field stores information about the key" rather than the document. I think we can also explicitly say that the driver:

1.  creates the key
2. inserts a new document into the keyvault with info about the key
3. returns the _id of the inserted document

File: source/use-cases/sensitive-data-encryption.txt [hunk 5]
```diff
@@ -124,7 +124,153 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+* The data encryption key, encrypted by the master key.
+* The key management service provider data.
+
+Follow the steps below to generate a single data encryption key from the
+local master key.
+
+.. note::
+
+  If using AWS KMS to provide the master key, refer to the "`F. Convert to
+  a Remote Master Key`_" section for the appropriate settings.
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: nit: the "F" in the step title is a bit weird. Instead we could add an inline reference above the step heading and then use a `:ref:` here instead

**Inline reference above the step heading**
```
.. _fle-convert-to-a-remote-master-key:

F. Convert to a Remote Master Key
```

**Reference in this note**
```
If using AWS KMS to provide the master key, refer to the :ref:`Convert
to a Remote Master Key <fle-convert-to-a-remote-master-key>` section of
this guide for the appropriate settings.
```

File: source/use-cases/sensitive-data-encryption.txt [hunk 6]
```diff
@@ -124,7 +124,153 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+
+First, read the file containing the local master key generated in the "`A.
+Create a Master Key`_" section:
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: See my previous comment

File: source/use-cases/sensitive-data-encryption.txt [hunk 7]
```diff
@@ -124,7 +124,153 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+
+          byte[] fileBytes = new byte[96];
+
+          try (FileInputStream fis = new FileInputStream(path)) {
+            fileBytes = fis.readAllBytes();
+          } catch (IOException e) {
+            e.printStackTrace();
+          }
+
+          final byte[] localMasterKey = Arrays.copyOf(fileBytes, 96);
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: - The code here looks misaligned though it doesn't seem to cause a rendering issue. The first character of each line of code should be even with the "c" in `code-block`.
- It'd be useful to highlight the line where we actually have the master key in memory. e.g.

```rst
.. code-block:: java
   :emphasize-lines: 11
   
   String path = "master-key.txt";
   // ... the rest of the code
```

File: source/use-cases/sensitive-data-encryption.txt [hunk 8]
```diff
@@ -124,7 +124,153 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+
+
+Next, specify the KMS provider settings. Since we are using a local master
+key, the provider name is ``local``:
+
+.. tabs::
+
+  tabs:
+
+    - id: java-kms-settings-local
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: All of our custom tab id values should be consistent, otherwise they won't stay in sync. (Eventually we should probably switch to pre-defined tabsets, but this is fine for now if we fix the id)

i.e. all the Java tabs should have the same ID, etc.

File: source/use-cases/sensitive-data-encryption.txt [hunk 9]
```diff
@@ -124,7 +124,153 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+            fileBytes = fis.readAllBytes();
+          } catch (IOException e) {
+            e.printStackTrace();
+          }
+
+          final byte[] localMasterKey = Arrays.copyOf(fileBytes, 96);
+
+
+Next, specify the KMS provider settings. Since we are using a local master
+key, the provider name is ``local``:
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: How do we specify the KMS provider settings? Looks like we make a hashmap in the code but why? What does it contain?

We should add a little bit more here to explain

File: source/use-cases/sensitive-data-encryption.txt [hunk 10]
```diff
@@ -124,7 +124,153 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+
+          Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>() {{
+             put(kmsProvider, new HashMap<String, Object>() {{
+                 put("key", localMasterKey);
+             }});
+          }};
+
+
+Create the client with the MongoDB connection string and key vault namespace
+(database and collection name):
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: Let's give specific guidance about what the namespace should be. If we have a best practice, we should say it. Otherwise, we should explicitly say that the names are up to the developer.

File: source/use-cases/sensitive-data-encryption.txt [hunk 11]
```diff
@@ -124,7 +124,153 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+    "creationDate" : ISODate("..."),
+    "updateDate" : ISODate("..."),
+    "status" : 0,
+    "masterKey" : {
+      "provider" : "local"
+    }
+  }
+
+`Click here <https://raw.githubusercontent.com/mongodb/docs-assets/DOCSP-csfle-data-encryption-key/DataEncryptionKeyGenerator.java>`_
+for the source code used to generate the data encryption key in this section.
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: nit: [Don’t provide links from ambiguous phrases such as Click here or More information](https://docs.mongodb.com/meta/style-guide/style/links-and-cross-references/)

File: source/use-cases/sensitive-data-encryption.txt (hunk 9 above)
Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: Also these "next..." phrases are really small relative to the code blocks they're sandwiched between. Given that this step is procedural, maybe we can convert to a stepfile?

File: source/use-cases/sensitive-data-encryption.txt (hunk 10 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Would it be ok if I just make a note to the user that is the arbitrary namespace we will be using for now? I discussed with Kenn and Ravind yesterday and the conclusion was that they would work on something more standardized during "beta cleanup". Conversation in #fle on Slack starts here:
https://mongodb.slack.com/archives/CDM9QEGUT/p1567530749015300

File: source/use-cases/sensitive-data-encryption.txt (hunk 11 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Thanks for pointing that out, will fix.

File: source/use-cases/sensitive-data-encryption.txt (hunk 8 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Oops. Up to this point, I thought they should be unique. Will need to fix across multiple PRs. I'll create a JIRA ticket to do cleanup on the guide once all the sections have been merged and add this as an item to address. For now, I will use `use-cases-fle-guide-java` (use-cases section, fle guide document, language)

File: source/use-cases/sensitive-data-encryption.txt (hunk 4 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: > Why.....
I'd imagine this has to do with the crypto library and someone's decision not to convert it to an ObjectId. It does not seem that important to the user, however, since it would only be defined in the JSON Schema.

Per the second comment, I will try to reformat as an ordered list.

File: source/use-cases/sensitive-data-encryption.txt (hunk 5 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Thanks, I was looking for that directive, but couldn't find it for some reason. Will fix.

File: source/use-cases/sensitive-data-encryption.txt (hunk 6 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Thanks, will fix.

File: source/use-cases/sensitive-data-encryption.txt (hunk 7 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Thanks, didn't know about the emphasize-lines directive. Will fix.

File: source/use-cases/sensitive-data-encryption.txt (hunk 9 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: > Looks like we make a hashmap in the code but why? What does it contain?
>
> We should add a little bit more here to explain

Can you explain what you're looking for? Each driver has its own way of specifying settings whether hashmap, json, associative array, etc. Did you want to have explicit information about how drivers specify their configurations? Would that belong in the scope of this document?

File: source/use-cases/sensitive-data-encryption.txt (hunk 8 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Will fix.

File: source/use-cases/sensitive-data-encryption.txt (hunk 3 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Agree, it's awkward. I see that you want to emphasize/include discussion of the key vault collection. Since it's the first two sentences, would your goal be met if key vault collection was only mentioned in the second one?

File: source/use-cases/sensitive-data-encryption.txt (hunk 9 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: > Given that this step is procedural, maybe we can convert to a stepfile?

Good idea, will move to a step file.

Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Question about tabs in a step file:
Should the id use the same naming scheme as the document? If so, then they are not reusable in different documents, if there is ever that need. Should we make note of that somewhere?

Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Another question -- if I have a `:ref:` link in the step file, will it resolve to the anchors set in the document that includes the step file?

File: source/use-cases/sensitive-data-encryption.txt (hunk 3 above)
Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: I think that's probably fine. The main reason I had it in both is to ensure that readers understand the keyvault is just a collection.

File: source/use-cases/sensitive-data-encryption.txt [hunk 12]
```diff
@@ -124,7 +126,49 @@ In this step, we generate a local master key to expedite setup of our developmen
           }
 
 
-B. Define a JSON Schema
+.. _fle-create-a-data-encryption-key:
+
+B. Create a Data Encryption Key
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In this section, we configure the MongoDB driver generate a data
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: configure the MongoDB driver _**and/to**_ generate a data key...

File: source/use-cases/sensitive-data-encryption.txt [hunk 13]
```diff
@@ -124,7 +126,49 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+encryption key using the master key. The data encryption key, used to 
+automatically encrypt and decrypt fields in a document, is stored in a MongoDB 
+collection called the *key vault collection*.
+
+The following configuration values are required by the client to generate a new
+data encryption key:
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: passive -> active

the client requires the following configuration values...

File: source/use-cases/sensitive-data-encryption.txt [hunk 14]
```diff
@@ -124,7 +126,49 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+* A MongoDB connection string that authenticates on a running server.
+* The key vault namespace (database and collection).
+
+In this guide, we use the ``encryption`` database and ``__keyvault`` 
+collection as our key vault collection.
+
+The MongoDB driver performs the following to generate the data encryption key:
+
+1.  Creates a data encryption key.
+#.  Inserts a new document into the key vault collection.
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: What's in the document? Looks like we give an example document in the final substep - maybe we can elaborate on what's in the document there?

File: source/use-cases/sensitive-data-encryption.txt [hunk 15]
```diff
@@ -124,7 +126,49 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+.. note::
+
+  If using AWS KMS to provide the master key, refer to the :ref:`Convert
+  to a Remote Master Key <fle-convert-to-a-remote-master-key>` section for
+  the appropriate settings.
+
+.. include:: /includes/steps/create-data-encryption-key.rst
+
+`Download the source code <https://raw.githubusercontent.com/mongodb/docs-assets/DOCSP-csfle-data-encryption-key/DataEncryptionKeyGenerator.java>`_
+used to generate the data encryption key in this section.
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: Seems like this might get lost here at the bottom of the step. Maybe we could combine with the "follow the steps" sentence above the stepfile?

Follow the steps below to generate a single data encryption key from the local master key. You can also find the complete code example on GitHub.

File: source/includes/steps-create-data-encryption-key.yaml [hunk 16]
```diff
@@ -0,0 +1,120 @@
+title: Read the local master key from a file
```
Comment (nlarew) [inline source/includes/steps-create-data-encryption-key.yaml]: nit: Let's use title case for step titles.

See [Enforce a Document Schema](https://docs.mongodb.com/stitch/mongodb/enforce-a-document-schema/) for an example

File: source/includes/steps-create-data-encryption-key.yaml [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+ref: read-local-master-key-from-file
+level: 4
+content: |
+  First, read the file containing the local master key, generated in the
+  :ref:`Create a Master Key <fle-create-a-master-key>` section:
```
Comment (nlarew) [inline source/includes/steps-create-data-encryption-key.yaml]: nit: the local master key _that you_ generated

File: source/includes/steps-create-data-encryption-key.yaml [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+              final byte[] localMasterKey = Arrays.copyOf(fileBytes, 96);
+
+
+---
+title: Specify KMS provider settings
+ref: specify-kms-provider-settings
+level: 4
+content: |
+  Next, specify the KMS provider settings. Set the provider name to ``local``
+  when using a local master key:
```
Comment (nlarew) [inline source/includes/steps-create-data-encryption-key.yaml]: I think we could have a little more detail on the what the KMS provider settings actually are / what they configure.

File: source/includes/steps-create-data-encryption-key.yaml [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+     permissions on the specified key vault namespace.
+
+
+---
+title: Verify the data encryption key was created
+ref: verify-data-key-created
+level: 4
+content: |
+  Verify that the key was created in your MongoDB replica set in the specified
+  key vault namespace. The document should resemble the following:
```
Comment (nlarew) [inline source/includes/steps-create-data-encryption-key.yaml]: It'd be nice to have a quick explanation of what the document actually is/contains/represents beyond just the example

File: source/includes/steps-create-data-encryption-key.yaml [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+
+              Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>() {{
+                put(kmsProvider, new HashMap<String, Object>() {{
+                  put("key", localMasterKey);
+                }});
+              }};
+
+
+---
+title: Construct the client
```
Comment (nlarew) [inline source/includes/steps-create-data-encryption-key.yaml]: The end state of this step isn't just a constructed client, it's actually a created key. Maybe the title here should be **Create the Data Key** and then in the text below we explain that the client creates and inserts the key.

File: source/use-cases/sensitive-data-encryption.txt (hunk 14 above)
Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: Thoughts on moving this process into the content of step three? It's most relevant there IMO.

File: source/use-cases/sensitive-data-encryption.txt (hunk 12 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Oops, good catch, will fix.

File: source/use-cases/sensitive-data-encryption.txt (hunk 13 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Thanks, will fix.

File: source/includes/steps-create-data-encryption-key.yaml (hunk 16 above)
Comment (ccho-mongodb) [reply source/includes/steps-create-data-encryption-key.yaml]: Will do, thanks for the example.

File: source/use-cases/sensitive-data-encryption.txt (hunk 15 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Sounds good, will update.

File: source/includes/steps-create-data-encryption-key.yaml (hunk 20 above)
Comment (ccho-mongodb) [reply source/includes/steps-create-data-encryption-key.yaml]: Good point, will fix.

File: source/includes/steps-create-data-encryption-key.yaml (hunk 17 above)
Comment (ccho-mongodb) [reply source/includes/steps-create-data-encryption-key.yaml]: Thanks, that sounds better, will fix. Just for my own understanding, is "generated in..." an example of a non-restrictive clause? And we prefer to use which/that over commas? Or am I completely wrong?

File: source/includes/steps-create-data-encryption-key.yaml (hunk 19 above)
Comment (ccho-mongodb) [reply source/includes/steps-create-data-encryption-key.yaml]: I removed that description from the beginning of the section because I did not think it provided very useful information. Which of the fields listed here https://github.com/mongodb/docs-ecosystem/pull/535/files#diff-54590a1d9915ddb3aea9c0163a7a6c8cR111 would you consider important? Though I did not directly refer to the names, I described the document contains _id, masterKey, and keyMaterial fields.
Should I also mention keyAltNames (which I have not attempted to generate)?

File: source/includes/steps-create-data-encryption-key.yaml (hunk 17 above)
Comment (ccho-mongodb) [reply source/includes/steps-create-data-encryption-key.yaml]: Whoops, lost my comment here (created a review instead of single comment, and then subsequently discarded the review). I prefer your version, but the question was just about whether "generated in the..." is an example of a non-restrictive clause and that we prefer that/which over commas, or whether it is not even a non-restrictive clause.

File: source/use-cases/sensitive-data-encryption.txt (hunk 14 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: I'm fine with moving it to step three, I was just trying to centralize "pre-" and "post-conditions" in the initial section. Would it also make sense to then move "The client requires the following..." on line 139 to steps 2 and 3 when they are used in the code examples?

File: source/includes/steps-create-data-encryption-key.yaml (hunk 18 above)
Comment (ccho-mongodb) [reply source/includes/steps-create-data-encryption-key.yaml]: I was planning to mention the AWS KMS-specific ones in the final section of the use-case guide, but are there other settings and are they important to mention?

File: source/includes/steps-create-data-encryption-key.yaml (hunk 19 above)
Comment (nlarew) [reply source/includes/steps-create-data-encryption-key.yaml]: I don't think we necessarily need to mention specific fields. Maybe just "includes the key (or KMS info) and metadata that describes when the key was created and last updated".

Comment (ccho-mongodb) [reply source/includes/steps-create-data-encryption-key.yaml]: Great, let me know if this makes sense:

https://github.com/mongodb/docs-ecosystem/pull/535/commits/58163994ae3c0dc20ee19ba73fbfe17a960b90db

File: source/includes/steps-create-data-encryption-key.yaml [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+  .. tabs::
+
+     tabs:
+
+       - id: use-cases-fle-guide-java
+         name: "Java"
+         content: |
+
+           .. code-block:: java
+              :emphasize-lines: 8,9,12
```
Comment (ccho-mongodb) [inline source/includes/steps-create-data-encryption-key.yaml]: Do you think I might be overusing the `emphasize-lines` directive?

File: source/includes/steps-create-data-encryption-key.yaml (hunk 17 above)
Comment (nlarew) [reply source/includes/steps-create-data-encryption-key.yaml]: AFAIK non-restrictive clause just means that the clause does not increase specificity. I think this might qualify under that definition but I'm not certain.

I think the main way we use it is in appositives, but I don't know if we use verb phrase appositives anywhere. I personally don't prefer this style because it feels like passive voice, i.e.:

"read the file containing the local master key (that was) generated in the Create a Master Key section (by whom?)."

File: source/includes/steps-create-data-encryption-key.yaml (hunk 18 above)
Comment (nlarew) [reply source/includes/steps-create-data-encryption-key.yaml]: Don't think there's any settings we don't have in this or the final step. I'm thinking more of "the client uses this settings map to access the master key"

File: source/includes/steps-create-data-encryption-key.yaml (hunk 21 above)
Comment (nlarew) [reply source/includes/steps-create-data-encryption-key.yaml]: Nope! Only 3 lines is fine and the highlighted lines here seem relevant.

File: source/use-cases/sensitive-data-encryption.txt [hunk 22]
```diff
@@ -124,7 +126,44 @@ In this step, we generate a local master key to expedite setup of our developmen
...
+#.  Inserts a new document into the key vault collection.
+#.  Returns the ``_id`` field of the inserted document as a `UUID <https://en.wikipedia.org/wiki/Universally_unique_identifier>`_.
+
+Follow the steps below to generate a single data encryption key from the
+local master key. You can also download the `complete code example on
+GitHub <https://raw.githubusercontent.com/mongodb/docs-assets/DOCSP-csfle-data-encryption-key/DataEncryptionKeyGenerator.java>`_.
+
+.. note::
+
+  If using AWS KMS to provide the master key, refer to the :ref:`Convert
```
Comment (ccho-mongodb) [inline source/use-cases/sensitive-data-encryption.txt]: Remove this

File: source/use-cases/sensitive-data-encryption.txt (hunk 14 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Delete this list

File: source/includes/steps-create-data-encryption-key.yaml (hunk 17 above)
Comment (ccho-mongodb) [reply source/includes/steps-create-data-encryption-key.yaml]: Thanks for the detailed explanation!

File: source/includes/steps-create-data-encryption-key.yaml (hunk 18 above)
Comment (ccho-mongodb) [reply source/includes/steps-create-data-encryption-key.yaml]: Makes sense, will add detail.
